### PR TITLE
MULTIARCH-4515: add new endpoint to receive initrd.addrsize 

### PR DESCRIPTION
**Pull request by AmadeusPodvratnik**
Add new parameter to return initrd.addrsize file overlayed by the current size of 
initrd.img and ignition.
This is only valid for s390x architecture and is necessary to support LPAR hypervisor (Classic and DPM).

This PR is the second required PR (First is https://github.com/openshift/assisted-image-service/pull/198).

This PR is for Assisted Installer, Agent-based Installer and HCP.
This new hypervisor support will be documented accordingly in the installation docs (https://issues.redhat.com/browse/MULTIARCH-4514)

## How was this code tested?
Running the initrd_test suite

## Assignees
@omertuc 

/cc @carbonin 
/cc 

## Links
[](https://issues.redhat.com/browse/MULTIARCH-4515)[https://issues.redhat.com/browse/MULTIARCH-4515](https://issues.redhat.com/browse/MULTIARCH-4515)

## Checklist

- [x] Title and description added to both, commit and PR
- [x] Relevant issues have been associated
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit tests (note that code changes require unit tests)

### CODE REVIEW

**Comment by AmadeusPodvratnik** [comment]
/ assign @omertuc

File: README.md
Comment (AmadeusPodvratnik) [reply README.md:127]: @paul-maidment Thank you for bringing this up. Maybe I bring some light into the thoughts from s390x community. Values like 0 and 1 are very common. In boot files like parm files values rd.neednet=1 or zfcp.allow_lun_scan=0 are used. Therefor I choose the value that way. But I understand that for users working with different architectures would get confused.

So what if we do it that way:
- change the parameter from 'addrsize' to 'file'
- value initrd.img or initrrd.addrsize with default initrd.img

Is this a feasible way to go?

File: internal/handlers/handlers_suite_test.go
Comment (omertuc) [inline internal/handlers/handlers_suite_test.go:23]: what do these bytes mean exactly?

File: internal/handlers/initrd.go
Comment (omertuc) [inline internal/handlers/initrd.go:44]:
```suggestion
		httpErrorf(w, http.StatusBadRequest, "'addrsize' parameter only valid for the s390x architecture")
```

File: internal/handlers/initrd.go
Comment (omertuc) [inline internal/handlers/initrd.go:87]:
```suggestion
	if addrsize == "1" {
```

File: README.md
Comment (omertuc) [inline README.md:127]: It should just be a boolean. true or false

Comment (omertuc) [reply README.md:127]: It should just be a boolean. true or false

Comment (omertuc) [reply README.md:127]: See below suggestion for my recommended phrasing. I mostly didn't like the use of "returned" since it's a too generic word and doesn't give much context of when and what

File: internal/handlers/initrd.go
Comment (omertuc) [inline internal/handlers/initrd.go:119]: Please move all the code my comment is replying to to its own separate, well-named function

File: internal/handlers/initrd.go
Comment (omertuc) [inline internal/handlers/initrd.go:108]: Variable name does not match standard golang style guidelines, please run linting

File: README.md
Comment (omertuc) [reply README.md:127]: >from s390x community. Values like 0 and 1 are very common

This is the semantics of our HTTP service API, the conventions of the s390x community / kernel boot arguments don't apply

File: internal/handlers/handlers_suite_test.go
Comment (AmadeusPodvratnik) [reply internal/handlers/handlers_suite_test.go:23]: I choose the values for testing purpose to make it easier to debug and verify. If you checking the ISO having the initrd.addrsize it contains the values based on the kernel.img and initrd.img (the loading address of the initrd.img depends on the size of the kernel alligned to MiB boundary to get the PSW for the initrd).

**Comment by omertuc** [comment]
We've had a voice conversation to clear up some of my confusions about this PR, we've agreed to move this to a separate endpoint altogether, rather than having a query parameter change the file that's being returned

**Comment by AmadeusPodvratnik** [comment]
@omertuc 
> We've had a voice conversation to clear up some of my confusions about this PR, we've agreed to move this to a separate endpoint altogether, rather than having a query parameter change the file that's being returned from initrd to addrsize

As discussed I added a new endpoint and changed the code accordingly. I will close all conversation and pls have a look at the new changes.

File: internal/handlers/handlers_suite_test.go
Comment (omertuc) [reply internal/handlers/handlers_suite_test.go:23]: Are they arbitrary though? if they have some sort of meaning/format, please use something like e.g.:

```go
bs := make([]byte, 8)
binary.LittleEndian.PutUint64(bs, uint64(64))
```

 To generate them

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:1]: This file is almost entirely a copy of `internal/handlers/initrd.go`, could we move out their common parts into something like `initrd_utils.go` or something?

Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go:1]: I created an initrd_utils.go and put the common parts into it. I did not create an explicit test.go because the tests would be done via initrd-test.go and initrd_addrsize.go.

If you want to have a initrd_utils_test.go than I will move the tests form initrd_test into it. 
Thank you.

Comment (omertuc) [reply internal/handlers/initrd_addrsize.go:1]: No need for a separate test file, thanks

File: internal/handlers/images_test.go
Comment (omertuc) [inline internal/handlers/images_test.go:96]: Are these commented out intentionally? Please uncomment or remove entirely

File: internal/handlers/initrd.go
Comment (omertuc) [inline internal/handlers/initrd.go:35]: unnecessary comment

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:51]: remove unnecessary comment

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:72]: remove unnecessary comment

File: README.md
Comment (omertuc) [inline README.md:136]:
```suggestion
Downloads the initrd.addrsize containing the psw of the initrd (8 bytes) and the size of the initrd (8 bytes). This is for s390x architecture only.
```

File: README.md
Comment (omertuc) [inline README.md:134]: I would like to emphasize how specific this is, so let's have it be part of the name

```suggestion
### `GET /images/{image_id}/s390x-initrd-addrsize`
```

Comment (omertuc) [reply README.md:134]: Or are there other IBM architectures which this might be relevant to in the future?

File: README.md
Comment (omertuc) [inline README.md:141]:
```suggestion
- `arch`: the base image cpu architecture (for this endpoint, this must be set to s390x)
```

File: internal/handlers/handlers_suite_test.go
Comment (omertuc) [inline internal/handlers/handlers_suite_test.go:36]: The variable is unnecessary

```suggestion
	Expect(os.WriteFile(filepath.Join(filesDir, "images/initrd.addrsize"), []byte{1, 2, 3, 4, 5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8}, 0600)).To(Succeed())
```

File: internal/handlers/images.go
Comment (omertuc) [inline internal/handlers/images.go:22]: Also here, let's rename to `s390xInitrdAddrsize`, etc

File: internal/handlers/images_test.go
Comment (omertuc) [inline internal/handlers/images_test.go:86]: Would this work as well?

```suggestion
		stubInitrdAddrsize := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			Expect(r.Method).To(Equal(http.MethodGet))
			Expect(r.URL.Path).To(Equal(fmt.Sprintf("/images/%s/initrd-addrsize", imageID)))
			http.ServeContent(w, r, "initrd.addrsize", time.Now(), strings.NewReader("initrdcontent"))
		})

		imageHandler := &ImageHandler{
			initrdAddrsize: stubInitrdAddrsize,
		}
```

File: internal/handlers/initrd_utils.go
Comment (omertuc) [inline internal/handlers/initrd_utils.go:16]:
```suggestion
```

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:69]:
```suggestion
	addrsizeBytes := new(bytes.Buffer)
```

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:75]: Why 8?

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:76]:
```suggestion
		Length: len(addrsizeBytes),
```

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [inline internal/handlers/initrd_addrsize.go:89]: Do we really need this overlaay? Can't we just serve the addrsize bytes buffer?

File: README.md
Comment (AmadeusPodvratnik) [reply README.md:134]: No yet ... so for the time being s390x only.

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go:89]: Yes ... because I need to take the initrd.addrsize from file and copy the size of initrd appended with the ignition into the last 8 byte and give return the new initrd.addrsize file back.

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [reply internal/handlers/initrd_addrsize.go:76]: Yep

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [reply internal/handlers/initrd_addrsize.go:89]: Ah I see. Isn't that just 8 bytes? Can't we just read them and append the size and serve the 16-byte buffer?

File: README.md
Comment (AmadeusPodvratnik) [reply README.md:136]: Done

File: README.md
Comment (AmadeusPodvratnik) [reply README.md:134]: Done

File: README.md
Comment (AmadeusPodvratnik) [reply README.md:141]: Done

File: internal/handlers/handlers_suite_test.go
Comment (AmadeusPodvratnik) [reply internal/handlers/handlers_suite_test.go:36]: Done

File: internal/handlers/images.go
Comment (AmadeusPodvratnik) [reply internal/handlers/images.go:22]: Done

File: internal/handlers/initrd.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd.go:35]: Done - removed

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go:51]: Done - removed

File: internal/handlers/initrd_utils.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_utils.go:16]: Done

File: internal/handlers/initrd_addrsize.go
Comment (omertuc) [reply internal/handlers/initrd_addrsize.go]: I'm not sure how that addresses my question. Why do a "live edit" of an existing file, when it's just 16 bytes? Read the first 8 bytes (the psw?), append the size of the new initrd, and serve that. The overlay is useful for editing huge files, it's over-engineered for 2 numbers

Comment (omertuc) [reply internal/handlers/initrd_addrsize.go]: Why is this conversation marked as resolved?

Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go]: Sorry ... closed accidentally. Overlay removed and return new buffer.

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go]: Done.

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go]: Done

**Comment by omertuc** [comment]
Waiting for @carbonin 's approval

**Review by carbonin — DISMISSED** [review]

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go:89]: io.MultiReader returns an io.Reader but http.ServeContent requires an io.ReadSeeker and conversation will require several lines of code. So there should be no benefit.

File: internal/handlers/initrd_addrsize_test.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize_test.go:137]: Remove this test case because endpoint is for s390x only.

File: internal/handlers/initrd_utils.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_utils.go:15]: Remove _utils.go and move to initrd.go.

File: internal/handlers/middleware.go
Comment (AmadeusPodvratnik) [reply internal/handlers/middleware.go:37]: Will remove for the time being because I'm not sure here.

File: internal/handlers/images_test.go
Comment (AmadeusPodvratnik) [reply internal/handlers/images_test.go:90]: Content is checked.

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go]: removed.

File: internal/handlers/initrd_addrsize.go
Comment (AmadeusPodvratnik) [reply internal/handlers/initrd_addrsize.go]: removed.

File: internal/handlers/images.go
Comment (AmadeusPodvratnik) [reply internal/handlers/images.go]: Done

File: README.md
Comment (AmadeusPodvratnik) [reply README.md]: Done

**Comment by carbonin** [comment]
@AmadeusPodvratnik @omertuc I fixed the rebase issue and squashed the commits with a more accurate message, can you both make sure everything looks right?

**Comment by carbonin** [comment]
It looks like everything is, in fact, not okay.

@AmadeusPodvratnik did you perhaps forget to commit the change to add the `initrdOverlayReader` function back to the `initrd.go` file?

**Pull request closed without merging**
